src/wallet/hdkeystore.cpp

```diff
@@ -0,0 +1,256 @@
+        }
+        else if (fragment == "c")
+            return false;
+        else
+        {
+            CExtKey childKey;
+            int32_t nIndex;
+            if (!ParseInt32(fragment,&nIndex))
+                return false;
+            parentKey.Derive(childKey, (harden ? 0x80000000 : 0)+nIndex);
+            parentKey = childKey;
+        }
+    }
+    extKeyOut = parentKey;
+    return true;
+}
+
+bool CHDKeyStore::DeriveKey(const HDChainID chainID, const std::string keypath, CKey& keyOut) const
+{
+    //this methode requires no locking
+    CExtKey extKeyOut;
+    if (!PrivateKeyDerivation(keypath, chainID, extKeyOut))
+        return false;
+
+    keyOut = extKeyOut.key;
+    return true;
+}
+
+bool CHDKeyStore::DeriveKeyAtIndex(const HDChainID chainID, CKey& keyOut, std::string& keypathOut, unsigned int nIndex, bool internal) const
+{
+    CHDChain hdChain;
+    if (!GetChain(chainID, hdChain))
+        return false;
+
+    if (nIndex >= 0x80000000)
+        throw std::runtime_error("CHDKeyStore::DeriveKeyAtIndex(): No more available keys!");
+
+    keypathOut = hdChain.keypathTemplate;
+    boost::replace_all(keypathOut, "c", itostr(internal)); //replace the chain switch index
+
+    keypathOut += "/"+itostr(nIndex)+"'"; //add hardened flag
+
+    CExtKey extKeyOut;
+    if (!PrivateKeyDerivation(keypathOut, chainID, extKeyOut))
+        throw std::runtime_error("CHDKeyStore::DeriveKeyAtIndex(): Private Key Derivation failed!");
+    keyOut = extKeyOut.key;
+
+    return true;
+}
+
+unsigned int CHDKeyStore::GetNextChildIndex(const HDChainID& chainID, bool internal)
+{
+    std::vector<unsigned int> vIndices;
+
+    {
+        LOCK(cs_KeyStore);
+
+        CHDChain hdChain;
+        if (!GetChain(chainID, hdChain))
+            return false;
+
+        std::string keypathBase = hdChain.keypathTemplate;
+        boost::replace_all(keypathBase, "c", itostr(internal)); //replace the chain switch index
+
+        //get next unused child index
+        for (std::map<CKeyID, CKeyMetadata>::iterator it = mapKeyMetadata.begin(); it != mapKeyMetadata.end(); ++it)
+        {
+            //skip non hd keys
+            if (it->second.keypath.size() == 0)
+                continue;
+
+            std::string keysBaseKeypath = it->second.keypath.substr(0, it->second.keypath.find_last_of("/"));
+            std::string childStr = it->second.keypath.substr(it->second.keypath.find_last_of("/") + 1);
+            boost::erase_all(childStr, "'");
+            int32_t nChild;
+            if(it->second.chainID == chainID &&
+               it->second.keypath.substr(0, it->second.keypath.find_last_of("/")) == keypathBase &&
+               ParseInt32(childStr, &nChild))
+            {
+                vIndices.push_back(nChild);
+            }
+        }
+    }
+
+    for (unsigned int i=0;i<0x80000000;i++)
+        if (std::find(vIndices.begin(), vIndices.end(), i) == vIndices.end())
+            return i;
+
+    return 0;
```

Comment:
(nit?) technically this only returns when you have 0x80000000 keys already... clearly not a worry, but if reached should probably barf